docs/proposal/20220712-garbage-collection.md

```diff
@@ -0,0 +1,338 @@
+
+### Non-Functional
+
+<a name="NFR8">NFR8.</a> CAPA MUST be able to easily add additional AWS resource cleanup in the future.
+
+<a name="NFR9">NFR9.</a> Unit tests MUST exist for new garbage collection code.
+
+<a name="NFR10">NFR10.</a> e2e tests MUST exist for the new garbage collection code for both unmanaged and managed clusters..
+
+### Implementation Details/Notes/Constraints
+
+#### Proposed Changes
+
+In the initial implementation of garbage collection if the feature is enabled all clusters will by default be marked for garbage collection. But we will supply a means to opt-out at time of creation as per [FR4](#FR4).
+
+The proposed changes will be documented depending on if the cluster is being created/updated or deleted.
+
+> NOTE: garbage collection will be experimental initial and will be enabled via a feature flag.
+
+##### Cluster Creation
+
+The following sequence diagram depicts what will happen when you create a new cluster with CAPI/CAPA with this change. The numbers will be referenced in the following descriptions.
+
+![gc creation](20220712-garbage-collection-create.svg)
+
+###### API Changes
+
+If the garbage collection feature has been enabled via the feature flag then the point that a user can mark a cluster as opting out of garbage collection ([FR4](#FR4)) is when they apply the yaml (1). This will be accomplished by annotating the **AWSCluster** or **AWSManagedControlPlane** with the `aws.cluster.x-k8s.io/external-resource-gc` annotation and setting its value to **false**.
+
+If the `aws.cluster.x-k8s.io/external-resource-gc` annotation is absent or its value is set to **true** then the CAPA created cluster will be garbage collected.
+
+This annotation name will be in a publicly exported package.
+
+###### Controller Changes
+
+The controllers for `AWSCluster` and `AWSManagedControlPlane` will be modified so that on creation of the controllers you can indicate that the garbage collection feature flag is enabled. In [main.go](../../main.go) we will look to see if the feature flag is enabled and pass this in when creating the instances of the 2 controllers.
+
+In the **reconcileNormal** of both controllers we will add logic at the beginning of reconciliation related to garbage collection. It will be after adding the existing finalizer (3) but before any AWS resources have been created (7,8). The logic will only be called if the garbage collection feature flag is enabled.
+
+The logic will be encapsulated in a new service and will have an **Reconcile** method which will be called from the controller (4).
+
+After the gc service **Reconcile** has been called the normal provisioning of the cluster will occur (7,8,9,10).
+
+###### New Garbage Collection Service
+
+The logic of the actions to take when a cluster is created or deleted will be contained in a new garbage collection (gc_service) service.
```

Comment:
Is this a repeat of the previous sentence?

Reply:
Deleted as it was basically a repeat.